### PR TITLE
apply constexpr and if constexpr when possible

### PR DESCRIPTION
**Pull request by cyyever**
Now that we have full C++17 support, we can use if constexpr in some identified cases.

### <samp>🤖 Generated by Copilot at df4c16d</samp>

The pull request improves the performance, readability, and consistency of various function templates in the `ATen` and `torch` modules by using `constexpr` keywords and C++17 features. It also fixes some type conversion and overflow issues for different input and output types. The changes affect the code for distributions, BLAS, batch normalization, embedding bag, random number generation, vectorized operations, cuBLAS, XNNPACK, CUTLASS, and shape inference. The affected files include `DistributionsHelper.h`, `vec256_int.h`, `vec512_int.h`, `BlasKernel.cpp`, `IndexKernel.cpp`, `EmbeddingBag.cpp`, `Normalization.cpp`, `rng_test.h`, `vec_test_all_types.h`, `TransformationHelper.h`, `CUDABlas.cpp`, `DistributionKernels.cpp`, `DistributionTemplates.h`, `RangeFactories.cu`, `RangeFactories.cpp`, `qconv.cpp`, `StructuredSparseLinearCUTLASS.cu`, `vec_test_all_types.cpp`, and `shape_inference.cpp`.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
@cyyever Looks like there are some real failures as a result of Visual Studio limitations with constexpr.

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your PR.  I think it breaks Windows CUDA build with a landrace https://hud.pytorch.org/pytorch/pytorch/commit/461c03a93c0ac85837c1ef11afc0ec1dc8900d0c' -c landrace

```
error: For this host platform/dialect, an extended lambda cannot be defined inside the 'if' or 'else' block of a constexpr if statement
```

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 1]
```diff
@@ -326,7 +326,7 @@ void random_from_to_kernel(TensorIteratorBase& iter, uint64_t range, int64_t bas
 template<typename RNG>
 void random_full_64_bits_range_kernel(TensorIteratorBase& iter, RNG gen) {
   AT_DISPATCH_ALL_TYPES_AND(at::ScalarType::BFloat16, iter.dtype(), "random_full_64_bits_range_kernel_cuda", [&] {
-    if constexpr (std::is_same<scalar_t, int64_t>::value ||
+    if (std::is_same<scalar_t, int64_t>::value ||
```
Comment (Skylion007) [inline aten/src/ATen/native/cuda/DistributionTemplates.h]: @cyyever You should use the `CONSTEXPR_EXCEPT_WIN_CUDA` macro instead.

**Comment by Skylion007** [comment]
> @pytorchbot revert -m 'Sorry for reverting your PR. I think it breaks Windows CUDA build with a landrace https://hud.pytorch.org/pytorch/pytorch/commit/461c03a93c0ac85837c1ef11afc0ec1dc8900d0c' -c landrace
> 
> ```
> error: For this host platform/dialect, an extended lambda cannot be defined inside the 'if' or 'else' block of a constexpr if statement
> ```

This is odd because I saw this error on CI on trunk, but when it automatically reran the job it passed.

**Comment by huydhn** [comment]
> > @pytorchbot revert -m 'Sorry for reverting your PR. I think it breaks Windows CUDA build with a landrace https://hud.pytorch.org/pytorch/pytorch/commit/461c03a93c0ac85837c1ef11afc0ec1dc8900d0c' -c landrace
> > ```
> > error: For this host platform/dialect, an extended lambda cannot be defined inside the 'if' or 'else' block of a constexpr if statement
> > ```
> 
> This is odd because I saw this error on CI on trunk, but when it automatically reran the job it passed.

Yeah, it's odd.  I will rebase and reland this if the issue doesn't occur again.  I suspect that this is a flaky issue due to either something on Windows runners or sccache

**Comment by cyyever** [comment]
Windows + CUDA is not 'if constexpr' friendly...

**Review by malfet — CHANGES_REQUESTED** [review]
Perhaps I'm reading this PR wrong, but current diff removes contrexprs:
image

Also, since you are doing c++17 updates, please replace `std::is_same<X,Y>::value` with `std::is_same_v<X, Y>`

**Comment by cyyever** [comment]
@malfet Commit history was recovered and changes to XXX_v  were applied.

**Review by malfet — APPROVED** [review]
